expat/lib/expat.h

```diff
@@ -124,9 +124,36 @@ enum XML_Error {
   XML_ERROR_RESERVED_PREFIX_XMLNS,
   XML_ERROR_RESERVED_NAMESPACE_URI,
   /* Added in 2.2.1. */
-  XML_ERROR_INVALID_ARGUMENT
+  XML_ERROR_INVALID_ARGUMENT,
+  /* Added in 2.3.0 */
+  XML_ERROR_ENTITY_TOO_LARGE,
+  XML_ERROR_ENTITY_EXPANSION_LIMIT,
+  XML_ERROR_ENTITY_RECURSION_LIMIT
 };
 
+enum XML_Options {
+  /* Added in 2.3.0 */
+  XML_OPTION_NONE = 0,
+  XML_OPTION_HUGE_ENTITES = 1<<0 /* No limits for entity expansion */
+};
+
+/* Entity expansion protection
+ *
+ * Mitigation against billion laugh and quadratic blowup attacks.
+ *
+ * XML_ENTITY_RECURSION_LIMIT confines nesting of entities within entities.
+ * XML_ENTITY_EXPANSION_SIZE restricts the maximum length of entities.
+ * XML_ENTITY_EXPANSION_RATIO constrains the ratio between position in XML
+ * document and total amount of all expanded entities.
+ *
+ * The limits are modelled after libxml2's limits
+ * https://github.com/GNOME/libxml2/blob/v2.9.8/parser.c#L99
+ */
+#define XML_ENTITY_RECURSION_LIMIT 3
+#define XML_ENTITY_EXPANSION_SIZE 1023
+#define XML_ENTITY_EXPANSION_RATIO 10
```

Comment:
(Just to have said it somewhere: We will need to support adjustment of those defaults at runtime at some point, maybe even before merging to master. I'm all in with getting other things polished first, though.)

Reply:
No, I wouldn't go there. I made the settings flexible in ``defusedexpat``. In retrospective it was unnecessary API bloat. Even ``libxml2`` doens't allow to tune the limits. It has only an off-switch for limitation. If ``libxml2`` doesn't need extra options, then ``expat`` surely doesn't need them at all.

Reply:
Interesting thought. Let me think about it more.